ledger-core/insolar/payload/validation.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package payload
+
+import (
+	"github.com/insolar/assured-ledger/ledger-core/vanilla/throw"
+)
+
+type Validate interface {
+	Validate(currPulse PulseNumber) error
+}
+
+var _ Validate = &VStateReport{}
+
+func (m *VStateReport) Validate(currPulse PulseNumber) error {
+	if err := validateUnimplemented(m); err != nil {
+		return err
+	}
+
+	if m.GetAsOf().IsUnknown() {
+		return throw.New("AsOf can not be Unknown")
+	}
+
+	if m.GetAsOf() >= currPulse {
+		return throw.New("AsOf should be less that current pulse")
+	}
+
+	if m.GetObject().IsEmpty() {
+		return throw.New("Object should not be empty")
+	}
+
+	if !m.GetObject().IsSelfScope() {
+		return throw.New("Object reference should be selfScore")
+	}
+
+	if m.GetObject().GetLocal().Pulse() >= currPulse {
+		return throw.New("Object pulse should be less that current pulse")
+	}
+
+	switch m.GetStatus() {
+	case Unknown:
+		return throw.New("Unexepected Unknown status received")
+	case Ready:
+		if err := validateReadyStatus(m, currPulse); err != nil {
+			return err
+		}
+	case Empty:
+		if err := validateEmptyStatus(m, currPulse); err != nil {
+			return err
+		}
+	case Missing:
+	case Inactive:
+		if err := validateZeroPending(m); err != nil {
+			return err
+		}
+
+		if err := validateEmptyState(m); err != nil {
+			return err
+		}
+
+		if err := validateEmptyLatest(m); err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func validateEmptyStatus(m *VStateReport, currPulse PulseNumber) error {
+	if m.GetOrderedPendingCount() != 1 {
+		return throw.New("Should be one ordered pending")
+	}
+
+	if m.GetUnorderedPendingCount() != 0 {
+		return throw.New("Unordered pending count should be 0")
+	}
+
+	if !m.GetUnorderedPendingEarliestPulse().IsUnknown() {
+		return throw.New("Unordered pending earliest pulse should be Unknown")
+	}
+
+	objectPulse := m.GetAsOf()
+	orderedPendingPulse := m.GetOrderedPendingEarliestPulse()
+
+	if orderedPendingPulse < objectPulse || currPulse < orderedPendingPulse {
+		return throw.New("Incorrect pending ordered pulse number")
+	}
+
+	return nil
+}
+
+func validateReadyStatus(m *VStateReport, currPulse PulseNumber) error {
+	if m.GetUnorderedPendingCount() < 0 || m.GetUnorderedPendingCount() >= 127 {
+		return throw.New("UnorderedPendingCount should be in range [0..127)")
+	}
+
+	if m.GetOrderedPendingCount() < 0 || m.GetOrderedPendingCount() >= 127 {
+		return throw.New("OrderedPendingCount should be in range [0..127)")
+	}
+
+	if m.GetUnorderedPendingEarliestPulse() < 0 || m.GetUnorderedPendingEarliestPulse() > currPulse {
```

Comment:
Not zero, https://github.com/insolar/assured-ledger/blob/master/ledger-core/pulse/pulse_number.go#L15 have a look here. I think it should be 65536 (pulse.LocalRelative)